### ISSUE TITLE
JSON Editor, Dynamic Expressions, and Routines

### ISSUE THREAD

**Issue by 96LawDawg** [ArnoldSmith86/virtualtabletop#487]
When you put a dynamic expression like `"var rotation = ${rotation} + 90"` after something like GET, SET, or other functions with curly brackets, the JSON editor, does not recognize that you might want to go back to curly brackets functions.  So whereas before, you could just keep entering function after function, now when you throw in dynamic expressions, you have to manipulate or trick the editor, or cut and paste or something in order to continue entering regular functions.

**Comment by ArnoldSmith86** [comment]
I'm guessing you don't indent your dynamic expression "correctly". Which is totally understandable and I agree that the JSON editor needs to receive a major update for dynamic expressions.

If you indent it like this it should work as before though:

```
  "clickRoutine": [
    {
      "func": "GET"
    },
    {
      "func": "DELETE"
    },
    {
      "func": "GET"
    },
    "var a = 1"
  ]
```

**Comment by 96LawDawg** [comment]
You are correct.  I was one space off.  I mean, I try to indent some because if you don't the "var" part will not get the right color.  Thanks for letting me know.  Easier to fix one space than cut and paste.

**Comment by RaphaelAlvez** [comment]
> I'm guessing you don't indent your dynamic expression "correctly". Which is totally understandable and I agree that the JSON editor needs to receive a major update for dynamic expressions.
> 
> If you indent it like this it should work as before though:
> 
> ```
>   "clickRoutine": [
>     {
>       "func": "GET"
>     },
>     {
>       "func": "DELETE"
>     },
>     {
>       "func": "GET"
>     },
>     "var a = 1"
>   ]
> ```

I just opened GitHub just to open an issue requesting auto-indentation. What a coincidencece

**Comment by ArnoldSmith86** [comment]
#488 should make this a lot easier to use.

After #469 my plan would be to divide the compute functions into groups (like `string`, `array`, ...), then have one button per group to add a new expression and when an expression of group `string` is active it shows commands for all `string` functions (to change the currently highlighted expression).

In time we can add a lot of other helpful tools but I don't have much time ATM and want to focus on #95 and then #479.

**Comment by ArnoldSmith86** [comment]
I marked this to be fixed by #557 because it adds automatic indentation which should make this way less of an issue.

If you feel like that's not enough, please unlink it or reopen it once it's closed.
